crates/red_knot_test/src/parser.rs

```diff
@@ -269,57 +252,143 @@ impl<'s> Parser<'s> {
+        const CODE_BLOCK_END: &[u8] = b"```";
+
+        while !self.cursor.is_eof() {
+            self.skip_whitespace();
+
+            match self.cursor.first() {
+                '#' => {
+                    self.explicit_path = None;
+                    self.preceding_blank_lines = 0;
+
+                    // Determine header level (number of '#' characters)
+                    let mut header_level = 0;
+                    while self.cursor.eat_char('#') {
+                        header_level += 1;
+                    }
+
+                    // Parse header title
+                    if let Some(title) = self.consume_until(|c| c == '\n') {
+                        let title = title.trim();
+
+                        if !title.is_empty() {
+                            self.process_header(header_level, title)?;
                         }
                     }
-                    '`' => {
-                        if let Some(captures) = CODE_RE.captures(self.cursor.as_str()) {
-                            self.parse_code_block(&captures)?;
-                            self.cursor.skip_bytes(captures.get(0).unwrap().len());
-                            continue;
+                }
+                '`' => {
+                    self.cursor.eat_char('`');
+
+                    match self.cursor.first() {
+                        '`' if self.cursor.second() == '`' => {
+                            // We see the triple-backtick beginning of a code block.
+                            self.cursor.bump();
+                            self.cursor.bump();
+
+                            if self.preceding_blank_lines < 1 && self.explicit_path.is_none() {
+                                bail!("Code blocks must start on a new line and be preceded by at least one blank line.");
+                            }
+
+                            self.skip_whitespace();
+
+                            // Parse the code block language specifier
+                            let lang = self
+                                .consume_until(|c| matches!(c, ' ' | '\n'))
+                                .unwrap_or_default();
+
+                            self.skip_whitespace();
+
+                            if self.cursor.first() != '\n' {
+                                bail!("Trailing code-block metadata is not supported. Only the code block language can be specified.");
+                            }
+
+                            self.skip_to_beginning_of_next_line();
+
+                            if let Some(position) =
+                                memchr::memmem::find(self.cursor.as_bytes(), CODE_BLOCK_END)
+                            {
+                                let mut code = &self.cursor.as_str()[..position];
+                                self.cursor.skip_bytes(position + CODE_BLOCK_END.len());
+
+                                if code.ends_with('\n') {
+                                    code = &code[..code.len() - 1];
```

Comment:
Nit: I find using `char::len_utf8` improves readability of code handling byte offsets because it makes it explicit what character we're offseting

```suggestion
                                    code = &code[..code.len() - '\n'.len_utf8()];
```